packages/redis/src/RedisStreamSource.ts

```diff
@@ -0,0 +1,233 @@
+        this.tracer = context.tracer;
+
+        this.client = makeLifecycle(redisClient(this.config));
+        await this.client.initialize(context);
+
+        const span = this.tracer.startSpan(this.spanOperationName);
+
+        this.spanLogAndSetTags(
+            span,
+            this.config.db,
+            this.config.readStream,
+            this.config.consumerGroup
+        );
+
+        try {
+            // Attempt to create stream + consumer group if they don't already exist
+            await this.client.xGroup(
+                span.context(),
+                this.config.readStream,
+                this.config.consumerGroup,
+                this.config.consumerGroupStartId
+            );
+        } catch (error) {
+            failSpan(span, error);
+            throw error;
+        } finally {
+            span.finish();
+        }
+    }
+
+    public async dispose(): Promise<void> {
+        await this.client.dispose();
+    }
+
+    private async getPendingMessagesForConsumer(): Promise<IRedisMessage[]> {
+        const span = this.tracer.startSpan(this.spanOperationName);
+
+        this.spanLogAndSetTags(
+            span,
+            this.config.db,
+            this.config.readStream,
+            this.config.consumerGroup
+        );
+        try {
+            const messages = await this.client.xReadGroup(
+                span.context(),
+                this.config.readStream,
+                this.config.consumerGroup,
+                this.config.consumerId,
+                0, // this will retrieve all PEL messages for this consumer
+                "0"
+            );
+
+            return messages;
+        } catch (error) {
+            failSpan(span, error);
+            throw error;
+        } finally {
+            span.finish();
+        }
+    }
+
+    private async getPendingMessagesForConsumerGroup(): Promise<IRedisMessage[]> {
+        const span = this.tracer.startSpan(this.spanOperationName);
+
+        this.spanLogAndSetTags(
+            span,
+            this.config.db,
+            this.config.readStream,
+            this.config.consumerGroup
+        );
+        try {
+            const pendingMessages = await this.client.xPending(
+                span.context(),
+                this.config.readStream,
+                this.config.consumerGroup,
+                this.config.batchSize
+            );
+
+            // if there are no pending messages return early w/ an empty array
+            if (pendingMessages.length < 1) {
+                return [];
+            }
+
+            const pendingMessagesIds = pendingMessages.map(({ streamId }) => streamId);
+
+            const messages = await this.client.xClaim(
+                span.context(),
+                this.config.readStream,
+                this.config.consumerGroup,
+                this.config.consumerId,
+                this.config.idleTimeoutMs,
```

Comment:
just for my understanding: `idleTimeoutMs` defines how long the server will block the call to wait for new msgs in case there are none when it receives the call?

Reply:
No `idleTimeoutMs` is the period a message has to sit on the PEL before it can be claimed by another consumer

Reply:
k, then I think there might be two more things we need:

1.) some kind of idle back off when it comes back with no new messages ... otherwise we are running a hot poll loop against redis on quiet streams.
2.) we probably don't want to call `getPendingMessagesForConsumerGroup` on every iteration, only occasionally

Reply:
Aren't both of these asks essentially the same thing? And I'm wondering if the added complexity of adding it/them is worth it considering the cost of the actual operations.

Reply:
I guess the idle timeout is for `getNewMessages` and `getPendingMessagesForConsumerGroup ` ... i would guess the network round trip time for these calls on empty streams is <1ms ... if we dont do any back-off we would query redis 1,000 times per second. so if neither of those two calls returns new messages then we should `await sleep(configurableAmountOfTime)` before doing the next iteration

Reply:
Ahhh I see your concern, that won't happen because `getNewMessages` is blocking.  I.e. it will block on `await` until new Messages enter the stream effectively preventing the above situation from occuring.

Reply:
https://github.com/dmmulroy/cookie-cutter/pull/2/files#diff-1025048fcd099a7a150de5b63b201456R378-R379

Reply:
i see, then I got another concern though :-) if there are no new messages then the application won't be able to shutdown gracefully. is there a way to specify a timeout for the blocking call?

Reply:
Yep I can add that as a config option 👍

Reply:
Updated, check it out.